src/imgui/ImGuiMachine.cc

```diff
@@ -46,8 +46,69 @@ ImGuiMachine::ImGuiMachine(ImGuiManager& manager_)
 	, setupFileList("setup", Reactor::SETUP_EXTENSION, Reactor::SETUP_DIR)
 	, confirmDialog("Confirm##setup")
 {
-	setupFileList.selectAction = [&](const FileListWidget::Entry& entry) {
-		manager.executeDelayed(makeTclList("setup", entry.displayName));
+
+	setupFileList.drawAction = [&] {
+		im::Table("table", 2, ImGuiTableFlags_BordersInnerV, [&]{
+			if (ImGui::TableNextColumn()) {
+				setupFileList.drawTable();
+			}
+			if (ImGui::TableNextColumn()) {
+				if (previewSetup.motherBoard) {
+					showSetupOverview(*previewSetup.motherBoard);
+					ImGui::Separator();
+					if (ImGui::Button("Run!")) {
+						manager.executeDelayed([&previewSetup = previewSetup, &manager = manager] {
+							manager.getReactor().switchMachineFromSetup(previewSetup.fullName);
+						});
+					}
+				} else {
+					if (previewSetup.lastExceptionMessage.empty()) {
+						ImGui::TextUnformatted("Nothing to preview...");
+					} else {
+						im::StyleColor(ImGuiCol_Text, getColor(imColor::ERROR), [&]{
+							ImGui::TextUnformatted(tmpStrCat("Setup ", previewSetup.displayName, " cannot be loaded:"));
+							ImGui::TextUnformatted(previewSetup.lastExceptionMessage);
+						});
+					}
+				}
+			}
+		});
+	};
+
+	setupFileList.displayColor = [&](const FileListWidget::Entry& entry) {
+		return !previewSetup.lastExceptionMessage.empty() && previewSetup.fullName == entry.fullName ? imColor::ERROR : imColor::TEXT;
+	};
+
+	setupFileList.singleClickAction = [&](const FileListWidget::Entry& entry) {
+		if (previewSetup.fullName == entry.fullName) return;
+
+		// record entry names, but (so far) without loaded motherboard
+		// this prevents that when loading failed, we don't continue retrying
+		previewSetup.fullName = entry.fullName;
+		previewSetup.displayName = entry.displayName;
+		// but we shouldn't reset the motherBoard yet during painting...
+
+		manager.executeDelayed([&previewSetup = previewSetup, &manager = manager] {
+			try {
+				// already reset, so that it's also gone in case of an exception
+				previewSetup.motherBoard.reset();
+				previewSetup.lastExceptionMessage.clear();
+				auto newBoard = manager.getReactor().createEmptyMotherBoard();
+				XmlInputArchive in(previewSetup.fullName);
+				in.serialize("machine", *newBoard);
+				previewSetup.motherBoard = newBoard;
+			} catch (MSXException& e) {
+				previewSetup.lastExceptionMessage = e.getMessage();
+			}
+		});
+	};
+
+	setupFileList.doubleClickAction = [&](const FileListWidget::Entry& entry) {
+		// only execute if there was no error when previewing this entry (if we did)
+		if (entry.fullName == previewSetup.fullName && !previewSetup.lastExceptionMessage.empty()) return;
+		manager.executeDelayed([&entry = entry, &manager = manager] {
+			manager.getReactor().switchMachineFromSetup(entry.fullName);
```

Comment:
In theory this could fail (throw an exception). Even though a few lines above we've checked that there was no `lastExceptionMessage`. That check was done earlier in time (seconds/minutes earlier). E.g. it could be that some required files have been removed since that check.
No need for extensive error handling, but we must make sure we don't crash when this happens (maybe a try-catch with empty catch block is sufficient).

Reply:
Good point, I forgot about that. Fixed.